### PR TITLE
Fix possible LOGICAL_ERROR in binary ariphmetics monotonicity. 

### PR DESCRIPTION
**Pull request by KochetovNicolai**
Fix invalid evaluation of binary monotonic function in KeyCondition.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `LOGICAL_ERROR` `Arguments of 'plus' have incorrect data types` which may happen in PK analysis (monotonicity check). Fix invalid PK analysis for monotonic binary functions with first constant argument.

### CODE REVIEW

File: src/Functions/FunctionBinaryArithmetic.h [hunk 1]
```diff
@@ -1782,30 +1782,32 @@ class FunctionBinaryArithmeticWithConstants : public FunctionBinaryArithmetic<Op
                 {
                     ColumnsWithTypeAndName columns_with_constant
                         = {{left.column->cloneResized(1), left.type, left.name},
-                           {right.type->createColumnConst(1, point), right.type, right.name}};
+                           {removeNullable(right.type)->createColumnConst(1, point), removeNullable(right.type), right.name}};
```
Comment (amosbird) [inline src/Functions/FunctionBinaryArithmetic.h:1785]: Maybe do `removeNullable(recursiveRemoveLowCardinality(right.type))`?

Comment (KochetovNicolai) [reply src/Functions/FunctionBinaryArithmetic.h:1785]: Yes, totally. Fuzzer found it as well :)

File: tests/queries/0_stateless/02461_mullable_pk_monotonicity_bug.sql [hunk 2]
```diff
@@ -1,6 +1,27 @@
 create table tab (x Nullable(UInt8)) engine = MergeTree order by x settings allow_nullable_key = 1, index_granularity = 2;
 insert into tab select number from numbers(4);
 
+set allow_suspicious_low_cardinality_types=1;
+set max_rows_to_read = 2;
+
+SELECT x + 1 FROM tab where plus(x, 1) <= 2 order by x;
+SELECT x + 1 FROM tab where plus(x, 1::Nullable(UInt8)) <= 2 order by x;
+SELECT x + 1 FROM tab where plus(x, 1::LowCardinality(UInt8)) <= 2 order by x;
```
Comment (amosbird) [inline tests/queries/0_stateless/02461_mullable_pk_monotonicity_bug.sql:9]: What will happen if we check `LowCardinality(Nullable(UInt8))`?

File: src/Functions/FunctionBinaryArithmetic.h [hunk 3]
```diff
@@ -1778,11 +1779,14 @@ class FunctionBinaryArithmeticWithConstants : public FunctionBinaryArithmetic<Op
             // const +|- variable
             if (left.column && isColumnConst(*left.column))
             {
+                auto left_type = removeLowCardinality(removeNullable(left.type));
```
Comment (amosbird) [inline src/Functions/FunctionBinaryArithmetic.h:1782]: It doesn't seem right to me, as LowCardinality cannot be inside Nullable.

**Comment by KochetovNicolai** [comment]
Finally AST fuzzers are green!

**Comment by alexey-milovidov** [comment]
@KochetovNicolai They ain't: https://github.com/ClickHouse/ClickHouse/issues/44814

**Pull request merged**